# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/webgpu/Texture.cpp
# Commit: dd10bf41fcea
# Full Hash: dd10bf41fcea6aa2f40d4528f4a7afc22b1f389e
# Author: Nicolas Silva <nsilva@mozilla.com>
# Date: 2024-04-05 21:43:02
# Regressor Bug: 1881518
# File Overlap Count: 1
# Description:
#   Bug 1881518 - Recycle indices again. r=webgpu-reviewers,bradwerth
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D202433
# ==============================================================================

diff -r b2cba928a7bb -r dd10bf41fcea dom/webgpu/Texture.cpp
--- a/dom/webgpu/Texture.cpp	Fri Apr 05 07:28:07 2024 +0000
+++ b/dom/webgpu/Texture.cpp	Fri Apr 05 07:42:05 2024 +0000
@@ -46,21 +46,21 @@
 }
 
 void Texture::Cleanup() {
-  if (!mParent) {
+  if (!mValid) {
+    return;
+  }
+  mValid = false;
+
+  auto bridge = mParent->GetBridge();
+  if (!bridge) {
     return;
   }
 
-  auto bridge = mParent->GetBridge();
-  if (bridge && bridge->IsOpen()) {
+  if (bridge->CanSend()) {
     bridge->SendTextureDrop(mId);
   }
 
-  // After cleanup is called, no other method should ever be called on the
-  // object so we don't have to null-check mParent in other places.
-  // This serves the purpose of preventing SendTextureDrop from happening
-  // twice. TODO: Does it matter for breaking cycles too? Cleanup is called
-  // by the macros that deal with cycle colleciton.
-  mParent = nullptr;
+  wgpu_client_free_texture_id(bridge->GetClient(), mId);
 }
 
 Texture::~Texture() { Cleanup(); }